### DISCUSSION TITLE
A very basic arduino program

### DISCUSSION THREAD

**Discussion by R4MP4G3RXD** [jcassel/OctoPrint-Siocontrol#9] [General]
Hello, I recently got your plugin working, here's a very basic cut down version of your example code that works on most arduinos. You could publish this for people with less programing knowhow. It uses all the digital pins as outputs and all the analog pins as inputs, I added a new restart function since that doesn't exist otherwise and I've removed web functionality. It still uses your TimeRelease header because it's a nice way of doing things.

https://pastebin.com/Qwbjgk4e

Cheers!

**Comment by jcassel** [comment]
@R4MP4G3RXD 
Thanks for the ino.. appreciate the work there to pare it down to something basic. I guess I should have created some releases as the earlier code was much closer to this in the ESPF12 version. 

Something I noted as I was looking at it was that you have:
`#define COMPATIBILITY "SIOPlugin 0.6.3"`

This should by design cause the PlugIn to raise an issue of incompatibility with the hardware. 
Something like this will show in the logs...  

```
2023-05-15 08:14:07,802 - octoprint.plugins.siocontrol - INFO - IO Reported Compatibility as:line='CP:SIOPlugin 0.6.3'
2023-05-15 08:14:07,802 - octoprint.plugins.siocontrol - INFO - Required Compatibility is:self.plugin.DeviceCompatibleVersion='SIOPlugin 0.1.1'
2023-05-15 08:14:07,802 - octoprint.plugins.siocontrol - INFO - cleared Command queue
2023-05-15 08:14:08,804 - octoprint.plugins.siocontrol - ERROR - IO Not compatible with this version of SIOPlugin
2023-05-15 08:14:08,804 - octoprint.plugins.siocontrol - ERROR - Stopping communications to SIO
2023-05-15 08:14:08,804 - octoprint.plugins.siocontrol - ERROR - Could not connect to SIO.
```

The COMPATIBILITY value is meant to only be adjusted when there is a breaking change to the PlugIn in relation to the Firmware.  Right now firmware compatibility it still 0.1.1 even though there have been a number of refinements in both the firmware and the PlugIn. In a perfect world, this value will never change. Reality is that it will once I add new capabilities like wireless IO connection to support virtualized instance. This is actually why you have stated to see "WiFi" capabilities in the examples for the ESPF12. I tend to work on that first for some reason. Anyway, I think that you have helped me realise just how much breaking changes might cause issues with existing systems that might not even be able to be updated to support the new capabilities.  I will put some extra thought and gating to ensure a high level of backward compatibility.  As well as adding some detailed context to updates that could require firmware changes.

**Reply by R4MP4G3RXD** [reply]
Ah yes, I've changed the version number after getting my program to work; A prime example of what haoppens if something that shouldn't be fixed gets fixed

**Comment by jcassel** [comment]
I created a minimal Arduino basic project here.  [SIO_Arduino_Minimal_Base](https://github.com/jcassel/OctoPrint_SIOControl_Firmware/tree/main/SIO_Arduino_Minimal_Base) 

I will keep this up to date along with the more advanced stuff where it makes sense. Likely it will not get all of the more advanced features but I will attempt to keep it in line as I add things like closed loop temp control.  Its really pretty hard for me to keep the memory foot print to a reasonable size and keep it reasonable to maintain.
